File: 1680-Concatenation of Consecutive Binary Numbers.py
class Solution(object):
    def concatenatedBinary(self, n):
        """
        :type n: int
        :rtype: int
        """
        kMod = 1000000007
        res = 1
        power = 1
        
        for i in range(2, n+1):
            if i&(i-1) == 0:
                power += 1
            res = ((res << power) | i)%kMod
        
        return res

File: C++/1680-Concatenation of Consecutive Binary Numbers.cpp
const int kMod = 1e9+7;
class Solution {
public:
    int concatenatedBinary(int n) {
        long long res = 0;
        for (int i=1; i<=n; i++) {
            int binaryLength = log2(i)+1;
            res = (res<<binaryLength)|i;
            res %= kMod;
        }

        return res;
    }
};


// Better
const int kMod = 1e9+7;
class Solution {
public:
    int concatenatedBinary(int n) {
        long long res = 0, binaryLength = 0;
        for (int i=1; i<=n; i++) {
            if ((i&(i-1)) == 0) binaryLength++;
            res = (res<<binaryLength)|i;
            res %= kMod;
        }

        return res;
    }
};